nixos/modules/services/networking/coredns.nix

```diff
@@ -0,0 +1,59 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.coredns;
+  configFile = pkgs.writeText "Corefile" cfg.config;
+in {
+  options.services.coredns = {
+    enable = mkEnableOption "Coredns dns server";
+
+    config = mkOption {
+      default = "";
+      example = ''
+        . {
+          whoami
+        }
+      '';
+      type = types.lines;
+      description = "Verbatim Corefile to use";
+    };
+
+    package = mkOption {
+      default = pkgs.coredns;
+      defaultText = "pkgs.coredns";
```

Comment:
FYI, I usually prefer to have such options only when you really need them or there's another good reason it should exist, but some people also like to have the option regardless. See https://github.com/NixOS/nixpkgs/issues/50476

Reply:
Here there is a good reason, coredns has a plugin system, these plugins should be enabled at compilation time by modifying the package